File: mvn-advent-of-code/src/main/java/advent/twenty_sixteen/Day04SecurityThroughObscurity.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Day04SecurityThroughObscurity implements DailyProblem<Integer, Integer> {
    private static final Comparator<Map.Entry<Character, Integer>> SORTED = (o1, o2) -> {
        if (o1.getValue() == o2.getValue())
            return Character.compare(o1.getKey(), o2.getKey());
        return Integer.compare(o2.getValue(), o1.getValue());
    };

    private final int part1Answer;
    private final int part2Answer;

    public Day04SecurityThroughObscurity(String filename) {
        List<String> lines = FileUtilities.readLines(filename, Parsers::TO_STRING);

        Map<String, Integer> decrypted = new HashMap<>();

        int count = 0;
        for (String line : lines) {
            count += checkAndDecrypt(line, decrypted);
        }

        this.part1Answer = count;
        this.part2Answer = decrypted.getOrDefault("northpole object storage", -1);
    }

    private int checkAndDecrypt(String line, Map<String, Integer> decrypted) {
        int index = line.lastIndexOf("-");
        String name = line.substring(0, index);
        int sector = Integer.parseInt(line.substring(index + 1, line.indexOf("[")));
        String checksum = line.substring(line.indexOf("[") + 1, line.indexOf("]"));
        String realName = "";

        Map<Character, Integer> counts = new HashMap<>();
        for (Character c : name.toCharArray()) {
            if (c != '-') {
                counts.put(c, counts.getOrDefault(c, 0) + 1);
                realName += (char) ('a' + ((c - 'a') + sector) % 26);
            } else {
                realName += " ";
            }
        }

        List<Map.Entry<Character, Integer>> sorted = counts.entrySet().stream().sorted(SORTED).collect(Collectors.toList());

        String calculatedChecksum = "";
        for (int i = 0; i != checksum.length(); i++) {
            calculatedChecksum += sorted.get(i).getKey();
        }

        if (checksum.equals(calculatedChecksum)) {
            decrypted.put(realName, sector);
            return sector;
        }

        return 0;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day04SecurityThroughObscurity.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

class Day04SecurityThroughObscurity(filename: String) extends DailyProblem[Int, Int] {

  private val input = FileUtilities.readFile(filename)

  private def sorter(lhs: (Char, Int), rhs: (Char, Int)): Boolean = {
    if (lhs._2.equals(rhs._2)) {
      lhs._1 < rhs._1
    } else {
      lhs._2 > rhs._2
    }
  }

  private def validate(room: String): Boolean = {
    val checkSum = room.take(room.lastIndexOf('-')).filter(c => c != '-').foldLeft(Map[Char, Int]().withDefaultValue(0))((acc, c) => acc + (c -> (acc(c) + 1))).toList.sortWith(sorter).take(5).map(x => x._1).mkString
    room.endsWith("[" + checkSum + "]")
  }

  private def decrypt(key: String, count: Int): String =  key.map(c => (c match {
      case '-' => ' '
      case _ => (((c.toInt - 'a'.toInt + count) % 26) + 'a'.toInt).toChar
    })).mkString

  private val getId = "-([0-9]*)\\[([a-z]*)\\]".r

  private def getId(room: String): Int = room.drop(room.lastIndexOf('-')) match {
    case getId(id, checksum) => id.toInt
  }

  private val valid = input.filter(x => validate(x))

  private val decrypted = valid.map(room => (getId(room), room.take(room.lastIndexOf('-')))).foldLeft(Map[String, Int]())((acc, pair) => acc + (decrypt(pair._2, pair._1) -> pair._1))

  override val part1Answer: Int = valid.map(room => getId(room)).sum
  override val part2Answer: Int = decrypted.withDefaultValue(-1)("northpole object storage")
}
